File: oop-alternatives/map-reduce-example.hs
-- If I want the sum of a list
sum :: (Num a) => [a] -> a
sum list = foldr (+) list

-- What about a product?
product :: (Num a) => [a] -> a
product list = foldr (*) list

-- What about the sum of squares?
sumOfSquares :: (Num a) => [a] -> a
sumOfSquares list = foldr (+) (map (^2) list)

File: oop-alternatives/map-reduce-example.cpp
template<typename T, typename Iterator>
T sum(Iterator start, Iterator end) 
{
    return std::accumulate(start, end, T(0), std::plus<T>());
}

template<typename T, typename Iterator>
T product(Iterator start, Iterator end)
{
    return std::accumulate(start, end, T(0), std::times<T>());
}

template<typename T, typename Iterator>
T sumOfSquares(Iterator start, Iterator end)
{
    return std::accumulate(start, end, T(0), [](T x, T y) { return x*x + y*y; });
}
